### PR TITLE
pebble: fix compilation failure on 32bits machine

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR mainly fixes a few compilation failures on 32bits machine.

The most important one is field alignment on 32bits machine. According the the go issue https://golang.org/pkg/sync/atomic/#pkg-note-BUG. `On 32 bit platforms, only 64-bit aligned fields can be atomic.`
But in the pebble codebase, uint64/int64 atomic fields are not aligned properly which will lead to crash.

### CODE REVIEW

**Review by petermattis — COMMENTED** [review]
Looks good. Some minor suggestions below. As I mentioned on the issue, it would be good to figure out how to automatically catch these alignment violations. Perhaps by running CI on some 32-bit platform. Not sure that is possible with Travis.

> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/929#-:-MHfk2np7G41i05VRq-M:bwsn973)* status: 5 of 6 files reviewed, 4 unresolved discussions (waiting on @petermattis and @rjl493456442)

___
*[commit.go, line 210 at r3](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHfjNM2BqcFkEogb83h:-MHfjNM2BqcFkEogb83i:bz5vhp2) ([raw file](https://github.com/cockroachdb/pebble/blob/0c5275c3b45251a702a9c73d8a281cb498337c1b/commit.go#L210)):*
> ```Go
> // batch in the queue. See commitPipeline.publish for additional commentary.
> type commitPipeline struct {
> 	// Queue of pending batches to commit.
> ```

Can you add a comment about why this field needs to appear first (i.e. to ensure alignment of the fields it contains)?
___
*[commit.go, line 213 at r4](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHfjcfB5qhG9tptrjXF:-MHfjcfB5qhG9tptrjXG:b-opdnxi) ([raw file](https://github.com/cockroachdb/pebble/blob/e1c54454dd2f145210fb65af3073d61e0deac4fe/commit.go#L213)):*
> ```Go
> 	pending commitQueue
> 	env     commitEnv
> 	sem     chan struct{}
> ```

This last commit feels like it should be merged with the previous commit. We generally try to make commits self-contained units.

PS Setting up `gofmt`-on-save in your editor eliminates needing to remember to do so. I'm only mentioning this because seeing a separate commit to `gofmt` makes me think you have don't so.
___
*[db.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHfiv5d8J43ZGI9xboG:-MHfiv5d8J43ZGI9xboH:b73k7aq) ([raw file](https://github.com/cockroachdb/pebble/blob/bbdc27104504fd3fe06200000e3524a2b3408087/db.go#L163)):*
> ```Go
> //	})
> type DB struct {
> 	// WARNING: The following fields are accessed atomically.
> ```

Nice comment! I think there is something else we can do to make it obvious that these fields are atomic: place them in a struct.

```
atomic struct {
  // ...
  memTableCount int64
  // ...
  memTableReserved int64
  // ...
  bytesFlushed uint64
  // ...
  bytesCompacted uint64
}
```

My hope is that by doing this, when we change this code in the future either the author or reviewer will notice if an atomic field is being added outside this struct.
___
*[go.mod, line 5 at r1](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHfihbh3a24l4UQJ6Fr:-MHfihbh3a24l4UQJ6Fs:bflctpl) ([raw file](https://github.com/cockroachdb/pebble/blob/51a87dae50e040184912a0e9debe42fa8b38cad7/go.mod#L5)):*
> ```XML
> require (
> 	github.com/certifi/gocertifi v0.0.0-20200211180108-c7c1fbc02894 // indirect
> 	github.com/cockroachdb/errors v1.7.5
> ```

Was the bump of `go.mod` needed for some reason? I prefer to keep such vendor bumps in a separate PR, or at least a separate commit.

**Comment by rjl493456442** [comment]
@petermattis All issues of accessing atomic fields have been fixed. Please take another look.

**Review by petermattis — COMMENTED** [review]
@rjl493456442 Generally looks good. A few nits and small comments below.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/929#-:-MHkiX_7ByNYPVGs0qu5:b-upyr9g)* status: 0 of 14 files reviewed, 10 unresolved discussions (waiting on @petermattis and @rjl493456442)

___
*[commit.go, line 210 at r5](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHkbJ3F-hrxfbYA0uYA:-MHkbJ3F-hrxfbYA0uYB:b9la5t5) ([raw file](https://github.com/cockroachdb/pebble/blob/eb069458d1373966a89f601315be1a5942d8a877/commit.go#L210)):*
> ```Go
> // batch in the queue. See commitPipeline.publish for additional commentary.
> type commitPipeline struct {
> 	// WARNING: The following struct `commitQueue` contains the field which will
> ```

Nit: `s/the field/fields/g`
___
*[db.go, line 163 at r5](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHkeZmU2E69qYCtF1-J:-MHkeZmU2E69qYCtF1-K:b5ui0vy) ([raw file](https://github.com/cockroachdb/pebble/blob/eb069458d1373966a89f601315be1a5942d8a877/db.go#L163)):*
> ```Go
> //	})
> type DB struct {
> 	// WARNING: The following struct `atomic` contains fields are accessed atomically.
> ```

Nit: `s/fields are/fields which are/g`
___
*[db.go, line 166 at r5](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHkeeUf7nCDbDQWWwcD:-MHkeeUf7nCDbDQWWwcE:b-gibx0c) ([raw file](https://github.com/cockroachdb/pebble/blob/eb069458d1373966a89f601315be1a5942d8a877/db.go#L166)):*
> ```Go
> 	//
> 	// On 32 bit platforms, only 64-bit aligned fields can be atomic. The struct is
> 	// guaranteed to be so aligned, so take advantage of that. For more information,
> ```

I'd rephrase the second sentence slightly: Go allocations are guaranteed to be 64-bit aligned which we take advantage of by placing the 64-bit fields which we access atomically at the beginning of the `DB` struct. I see you copied this comment to a few other places. Please update all of them.
___
*[db.go, line 187 at r5](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHkfYDK5cOLVd36y6JQ:-MHkfYDK5cOLVd36y6JR:bmbcdsr) ([raw file](https://github.com/cockroachdb/pebble/blob/eb069458d1373966a89f601315be1a5942d8a877/db.go#L187)):*
> ```Go
> 		// The size of the current log file (i.e. db.mu.log.queue[len(queue)-1].
> 		logSize uint64
> 	}
> ```

Nit: let's separate this struct from the following fields with a blank line.
___
*[db.go, line 251 at r5](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHkfeul6Q-1ACwrC-WH:-MHkfeul6Q-1ACwrC-WI:b-kxcezn) ([raw file](https://github.com/cockroachdb/pebble/blob/eb069458d1373966a89f601315be1a5942d8a877/db.go#L251)):*
> ```Go
> 		// The collection of immutable versions and state about the log and visible
> 		// sequence numbers.
> 		versions *versionSet
> ```

I think it is worth a comment explaining why this became a pointer in order to prevent someone in the future coming along and changing it back to a value.
___
*[table_cache.go, line 509 at r5](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHkgGuSAuMlFqV7hAby:-MHkgGuSAuMlFqV7hAbz:bf3voax) ([raw file](https://github.com/cockroachdb/pebble/blob/eb069458d1373966a89f601315be1a5942d8a877/table_cache.go#L509)):*
> ```Go
> func (c *tableCacheShard) Close() error {
> 	// The cache shard is not allocated yet, nothing to close
> 	if c == nil {
> ```

I think this check would be more natural in `tableCache.Close`, as that is the component which is allocating the `tableCacheShards`.

**Comment by rjl493456442** [comment]
@petermattis Updated, please take another look

**Review by petermattis — APPROVED** [review]
Thanks. Just a few more comment and style nits.

Does this fix all of the 64-bit atomic problems on 32-bit platforms? Or are there more remaining?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/929#-:-MHkt5wB84ylm3cycp09:b-eqdwht)* status: 0 of 14 files reviewed, 15 unresolved discussions (waiting on @petermattis and @rjl493456442)

___
*[commit.go, line 215 at r6](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHksUuP9QEAce4ng_rC:-MHksUuP9QEAce4ng_rD:b-81nkmr) ([raw file](https://github.com/cockroachdb/pebble/blob/46cf27183c00c1d0ca015d486df3b2f748814664/commit.go#L215)):*
> ```Go
> 	// Go allocations are guaranteed to be 64-bit aligned which we take advantage
> 	// of by placing the 64-bit fields which we access atomically at the beginning
> 	// of the DB struct. For more information, see https://golang.org/pkg/sync/atomic/#pkg-note-BUG.
> ```

Nit: `s/DB struct/commitPipeline struct/g`. Or perhaps just remove `DB` to make the comment general.
___
*[db.go, line 253 at r6](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHksyjGDUIECRF9QEU3:-MHksyjGDUIECRF9QEU4:bx1lg1a) ([raw file](https://github.com/cockroachdb/pebble/blob/46cf27183c00c1d0ca015d486df3b2f748814664/db.go#L253)):*
> ```Go
> 		// The collection of immutable versions and state about the log and visible
> 		// sequence numbers. Use the pointer here to ensure the atomic fields in
> 		// version set can be aligned properly.
> ```

Nit: `s/can be/are/g`
___
*[table_cache.go, line 122 at r6](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHksiAl5nw-P06iDmv0:-MHksiAl5nw-P06iDmv1:bioqk1o) ([raw file](https://github.com/cockroachdb/pebble/blob/46cf27183c00c1d0ca015d486df3b2f748814664/table_cache.go#L122)):*
> ```Go
> 		misses    int64
> 		iterCount int32
> 	}
> ```

Nit: blank line here as well.
___
*[version_set.go, line 43 at r6](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHkss3Y4IKsN2VH6kD-:-MHkss3Y4IKsN2VH6kD0:b-v14x8k) ([raw file](https://github.com/cockroachdb/pebble/blob/46cf27183c00c1d0ca015d486df3b2f748814664/version_set.go#L43)):*
> ```Go
> 	// Go allocations are guaranteed to be 64-bit aligned which we take advantage
> 	// of by placing the 64-bit fields which we access atomically at the beginning
> 	// of the DB struct. For more information, see https://golang.org/pkg/sync/atomic/#pkg-note-BUG.
> ```

Nit: `DB struct` is inaccurate here.
___
*[version_set.go, line 57 at r6](https://reviewable.io/reviews/cockroachdb/pebble/929#-MHksloP2jE0lNsxUvby:-MHksloP2jE0lNsxUvbz:bioqk1o) ([raw file](https://github.com/cockroachdb/pebble/blob/46cf27183c00c1d0ca015d486df3b2f748814664/version_set.go#L57)):*
> ```Go
> 		// compactions. Updated and read atomically.
> 		atomicInProgressBytes int64
> 	}
> ```

Nit: blank line here as well.

**Comment by rjl493456442** [comment]
@petermattis Done.

**Review by petermattis — COMMENTED** [review]
:lgtm: Thanks for the contibution, @rjl493456442. Does this fix all of the 64-bit atomic problems on 32-bit platforms? Or are there more remaining?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/929#-:-MHlMww-9mCS41gnAgvf:b-iblp4r)* status: 0 of 14 files reviewed, 1 unresolved discussion (waiting on @petermattis and @rjl493456442)

**Comment by rjl493456442** [comment]
@petermattis At least now I can run the pebble in my 32bits machine. So I would say this PR almost fixes all the issues. Maybe there are some "unused" package still contains the issue, but I think we can definitely fix them in a separate PR next time.

**Comment by SWDevAngel** [comment]
@rjl493456442 LAST DAY TO ORDER!!! 
Hi Gary. Thank you for contributing to CockroachDB this year. As a token of our appreciation we would like to send you a gift while supplies last. We tried to reach you via GitHub or on email, but if you didn’t see the message please DM me on our [community Slack](https://cockroa.ch/slack) `@lisa` so I can send you a link to order your gift. (If you don’t want a gift, we also have a charitable contribution choice.) All orders must be in by December 13 (that’s tomorrow!), so please send this asap. :)
